### PR TITLE
 issue #53: support multi vpc analysis

### PR DESCRIPTION
**Pull request by adisos**
Currently supporting this type of analysis with textual output formats (txt/md)

JSON and drawio not yet supported

### CODE REVIEW

File: pkg/ibmvpc/parser.go
Comment (zivnevo) [inline pkg/ibmvpc/parser.go:1]: A general comment: is there a guarantee for the uniqueness of VPC names?
I only found this sentence in the documentation: "The name must not be used by another VPC in the region.".
But we are collecting all VPCs of a given account, so maybe the same name is being used in two different regions?

File: pkg/ibmvpc/parser.go [hunk 1]
```diff
@@ -651,45 +684,46 @@ func getSGconfig(rc *ResourcesContainer,
...
 	skipByVPC func(string) bool) error {
 	// nacl
-	naclList := []*NACL{}
-	var layerVPC *VPC
-	for i := range rc.naclList {
-		nacl := rc.naclList[i]
+	naclList := map[string][]*NACL{} // map from vpc name to its nacls
+	for vpcnName := range res {
+		naclList[vpcnName] = []*NACL{}
+	}
```
Comment (zivnevo) [inline pkg/ibmvpc/parser.go:694]: Is this required? Appending to nil should work as well.

File: pkg/ibmvpc/parser.go [hunk 2]
```diff
@@ -651,45 +684,46 @@ func getSGconfig(rc *ResourcesContainer,
...
-	res *vpcmodel.CloudConfig,
+	res map[string]*vpcmodel.CloudConfig,
 	subnetNameToSubnet map[string]*Subnet,
 	skipByVPC func(string) bool) error {
 	// nacl
-	naclList := []*NACL{}
-	var layerVPC *VPC
-	for i := range rc.naclList {
-		nacl := rc.naclList[i]
+	naclList := map[string][]*NACL{} // map from vpc name to its nacls
```
Comment (zivnevo) [inline pkg/ibmvpc/parser.go:691]: rename `naclList` to `naclLists`

File: pkg/ibmvpc/parser.go [hunk 3]
```diff
@@ -557,90 +585,95 @@ func getFipConfig(
...
+	return nil
 }
 
-func singleVPCErr(vpcName, layerVPCName string) error {
-	return fmt.Errorf("NACL/SG VPC (%s) is different from its layer's VPC (%s)", vpcName, layerVPCName)
+func parseSGTargets(sgResource *SecurityGroup,
+	sg *vpc1.SecurityGroup,
+	res map[string]*vpcmodel.CloudConfig,
+	intfNameToIntf map[string]*NetworkInterface,
+	vpcName string) {
```
Comment (zivnevo) [inline pkg/ibmvpc/parser.go:613]: `res` and `vpcName` only appear as `res[vpcName]`.
Maybe only pass a pointer to the relevant `CloudConfig`.

File: pkg/ibmvpc/parser.go [hunk 4]
```diff
@@ -557,90 +585,95 @@ func getFipConfig(
...
-		} else if layerVPC.Name() != vpc.Name() {
-			return singleVPCErr(vpc.Name(), layerVPC.Name())
-		}
 
 		sgResource := &SecurityGroup{
 			VPCResource: vpcmodel.VPCResource{
 				ResourceName: *sg.Name, ResourceUID: *sg.CRN, ResourceType: ResourceTypeSG},
 			analyzer: NewSGAnalyzer(sg), members: map[string]vpcmodel.Node{}, vpc: vpc,
 		}
 		sgMap[*sg.Name] = sgResource
```
Comment (zivnevo) [inline pkg/ibmvpc/parser.go:662]: Isn't this problematic? Different Security Groups with the same name may live in different VPCs

**Review by zivnevo — CHANGES_REQUESTED** [review]
Some comments to start with.
I still haven't reviewed the fip code in `parser.go`.

File: cmd/analyzer/main.go [hunk 5]
```diff
@@ -51,35 +75,23 @@ func _main(cmdlineArgs []string) error {
...
 	}
 
-	rc, err := ibmvpc.ParseResourrcesFromFile(*inArgs.InputConfigFile)
+	rc, err := ibmvpc.ParseResourcesFromFile(*inArgs.InputConfigFile)
 	if err != nil {
 		return fmt.Errorf("error parsing input vpc resources file: %w", err)
 	}
 
-	cloudConfig, err := ibmvpc.NewCloudConfig(rc, *inArgs.VPC)
+	cloudConfig, err := ibmvpc.CloudConfigsFromResources(rc, *inArgs.VPC, *inArgs.Debug)
```
[the remark this answers is not in the document]
Comment (adisos) [reply cmd/analyzer/main.go]: Renamed

File: cmd/analyzer/main.go [hunk 6]
```diff
@@ -51,35 +75,23 @@ func _main(cmdlineArgs []string) error {
...
-	outFile := ""
-	if inArgs.OutputFile != nil {
-		outFile = *inArgs.OutputFile
-	}
-	outFormat := getOutputFormat(inArgs)
-	output, err := og.Generate(outFormat, outFile)
-	if err != nil {
-		return fmt.Errorf("output generation error: %w", err)
+	numConfigs := len(cloudConfig)
+	for configName, config := range cloudConfig {
```
[the remark this answers is not in the document]
Comment (adisos) [reply cmd/analyzer/main.go]: Added VPC name to this struct.
Kept the map, as it was needed for the internal implementation.

File: pkg/vpcmodel/output.go [hunk 7]
```diff
@@ -83,15 +84,36 @@ func (o *OutputGenerator) Generate(f OutFormat, outFile string) (string, error)
...
+// finalizeOutput adds info line about the name of the VPC before its analysis output,
+// in case multiple VPCs are analyzed separately
+func finalizeOutput(output string, numVPCs int, vpcName string, f OutFormat) string {
+	if numVPCs == 1 {
+		return output
+	}
+	res := output
+	switch f {
+	case Text, MD, Debug:
+		res = fmt.Sprintf("Analysis for VPC %s:\n%s", vpcName, output)
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/vpcmodel/output.go]: done

File: pkg/ibmvpc/parser.go [hunk 8]
```diff
@@ -909,12 +989,62 @@ func addExternalNodes(config *vpcmodel.CloudConfig, vpcInternalAddressRange *com
...
 		if res, ok := obj.(*VPC); ok {
 			return res, nil
 		}
 		return nil, fmt.Errorf("a resource of name %s is not a VPC as expected", vpcName)
 	}
-	return nil, fmt.Errorf("missing VPC resource of name %s", vpcName)
+	return nil, missingVPCErr
+}
+
+func printCloudConfigs(c map[string]*vpcmodel.CloudConfig) {
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/parser.go]: done

File: pkg/ibmvpc/parser.go [hunk 9]
```diff
@@ -557,90 +585,95 @@ func getFipConfig(
...
+			VPCResource: vpcmodel.VPCResource{ResourceName: vpcName, ResourceUID: *vpc.CRN, ResourceType: ResourceTypeVPC},
 			nodes:       []vpcmodel.Node{},
 			zones:       map[string]*Zone{},
 		}
-		res.NodeSets = append(res.NodeSets, vpcNodeSet)
-		res.NameToResource[vpcNodeSet.Name()] = vpcNodeSet
-		// assuming one vpc
-		return vpcNodeSet, nil
+		res[vpcName] = NewEmptyCloudConfig()
+		res[vpcName].NodeSets = append(res[vpcName].NodeSets, vpcNodeSet)
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/parser.go]: done

File: pkg/ibmvpc/parser.go
Comment (adisos) [reply pkg/ibmvpc/parser.go:1]: Changed to VPC unique IDs (CRN) instead of names.

File: pkg/ibmvpc/parser.go (hunk 1 above)
Comment (adisos) [reply pkg/ibmvpc/parser.go]: right, done.

File: pkg/ibmvpc/parser.go (hunk 2 above)
Comment (adisos) [reply pkg/ibmvpc/parser.go]: done

File: pkg/ibmvpc/parser.go (hunk 3 above)
Comment (adisos) [reply pkg/ibmvpc/parser.go]: done

File: pkg/ibmvpc/parser.go (hunk 4 above)
Comment (adisos) [reply pkg/ibmvpc/parser.go]: right, fixed (`sgMap` is now per each VPC)

File: pkg/ibmvpc/parser.go [hunk 10]
```diff
@@ -767,99 +814,145 @@ func getIKSnodesConfig(res *vpcmodel.CloudConfig,
...
 		NameToResource:   map[string]vpcmodel.VPCResourceIntf{},
 		CloudName:        "IBM Cloud",
 	}
+}
 
+// VPCConfigsFromResources returns a map from VPC UID (string) to its corresponding VPCConfig object,
+// containing the parsed resources in the relevant model objects
+func VPCConfigsFromResources(rc *ResourcesContainer, vpcID string, debug bool) (map[string]*vpcmodel.VPCConfig, error) {
+	res := map[string]*vpcmodel.VPCConfig{} // map from VPC UID to its config
+	vpcsMap := map[string]*VPC{}            // map from VPC UID to its object
```
Comment (zivnevo) [inline pkg/ibmvpc/parser.go:849]: Does it make sense to add to `VPCConfig` a pointer to the VPC resource (of type `VPCResourceIntf`)?
This way, we can maintain only one map (`res`) and get rid of `vpcsMap`.

File: pkg/ibmvpc/parser.go [hunk 11]
```diff
@@ -529,26 +536,51 @@ func getFipConfig(
...
-						}
+
+		if targetAddress == "" {
+			continue
+		}
+
+		var srcNodes []vpcmodel.Node
+		var vpcUID string
+		for uid, vpcConfig := range res {
+			srcNodes = getCertainNodes(vpcConfig.Nodes, func(n vpcmodel.Node) bool { return n.Cidr() == targetAddress })
```
Comment (zivnevo) [inline pkg/ibmvpc/parser.go:547]: Is this good enough in a multi-VPC scenario? Are we relying on address spaces to be disjoint?
Wouldn't it be better to compare the target's CRN?
If we do rely on disjoint address spaces, please add here a comment that this is the assumption.

**Review by zivnevo — CHANGES_REQUESTED** [review]
A few more comments

File: pkg/ibmvpc/examples/acl_testing3.md [hunk 12]
```diff
@@ -1,3 +1,4 @@
+Analysis for VPC test-vpc1-ky:
```
[the remark this answers is not in the document]
Comment (adisos) [reply pkg/ibmvpc/examples/acl_testing3.md]: done

File: pkg/ibmvpc/parser.go (hunk 11 above)
Comment (adisos) [reply pkg/ibmvpc/parser.go]: Changed to be based on unique id

File: pkg/ibmvpc/parser.go (hunk 10 above)
Comment (adisos) [reply pkg/ibmvpc/parser.go:849]: The `VPCConfig`  is defined in the level of the abstract model, and the `VPC` resource type is defined in the concrete `ibmvpc` pkg, where it is parsed with all other concrete resources.

**Review by zivnevo — APPROVED** [review]

File: pkg/ibmvpc/parser.go (hunk 10 above)
Comment (adisos) [reply pkg/ibmvpc/parser.go:849]: ok, will add as another item in the issue

**Pull request merged**